### DISCUSSION TITLE
Zertifikatsprobleme

### DISCUSSION THREAD

**Discussion by NEO1976** [DieTechniker/secon-tool#56] [Q&A]
Sehr geehrte Damen und Herren,

wir haben ein Problem mit Secon Tool. Wir haben ab und zu Kunden mit mehreren gültigen Zertifikate und wenn sie uns Daten schicken, passiert es oft, dass es in unserem System mit der falschen Zertifikatsversion von dem gleichen Kunden verschlüsselt ist. Wir bekommen leider keine Fehlermeldung von secon-tool, dass die Zertifikatssignaturen nicht passen. Deshalb kann der Kunde unsere verschlüsselte Rückmeldung nicht entschlüsseln. Gibt es einen Weg oder einen Befehl in secon-tool, damit wir die Signature prüfen können?

Mit freundlichen Grüßen

Hürol Türen

**Comment by loetifuss** [comment]
Hallo,

Seit dieser Änderung https://github.com/DieTechniker/secon-tool/pull/47 fügt secon-tool das für die Signatur verwendete Zertifikat in die Nachricht ein. Damit wird beim Empfänger die Signatur anhand dieses Zertifikats geprüft. Dadurch muss der Empfänger nur noch die Ausstellerzertifikate kennen.
Bei der Entschlüsselung wählt secon-tool den privaten Schlüssel aus, der zum [Zertifikat aus der Nachricht passt](https://github.com/DieTechniker/secon-tool/blob/master/src/main/java/de/tk/opensource/secon/DefaultSubscriber.java#L234). Wenn der Empfänger alle seine privaten Schlüssel kennt dürfte es auch mit mehreren gültigen Zertifikaten kein Problem geben. Oder nutzt der Empfänger ein anderes Tool für die Entschlüsselung?

**Reply by NEO1976** [reply]
Hallo,

Der Empfänger nutz kein anderes Tool für die Entschlüsselung.  Ich kann ein Beispiel geben:

Der Kunde A hat ein gültiges Zertifikat, dessen Enddatum zB 10.10.2023 ist. Der Kunde denkt, dass der Kunde sobald wie möglich ein neues Zertifikat von ITSG holen. Er hat ein neues Zertifikat, dessen Enddatum 10.10.2026. Dieser Kunde schickt uns eine Datei, die mit dem neuesten Zertifikat verschlüsselt ist. Wir entschlüsseln die Datei durch Secon.Tool. Secon Tool Software findet versehentlich das alte aber gültige Zertifikat und verschlüsselt die Rückmeldung mit dem falschen Zertifikat. Der Kunde kann deshalb die Rückmeldung nicht entschlüsseln.

Welche Version hat der Buld #47 ? Wir haben die Version 1.2.0. 

Mit freundlichen Grüßen

Hürol Türen

**Reply by loetifuss** [reply]
Um einen reibungslosen Betrieb zu gewährleisten sollte der Kunde alle gültigen privaten Schlüssel in seinem Keystore haben (in diesem Fall das neue und das alte Zertifikat). Es ist ja nicht sichergestellt, dass alle Kommunikationsteilnehmer gleichzeitig auf das neue Zertifikat umstellen. Damit ist der Empfänger in der Lage, alle Nachrichten zu entschlüsseln, egal welches Schlüsselpaar verwendet wurde.
